File: TranspositionCipherSolution.py
# Transposition Cipher Encryption
# https://www.nostarch.com/crackingcodes/ (BSD Licensed)

# import pyperclip
import math
import pprint


def encryptMessage(key, message):
    # Each string in ciphertext represents a column in the grid:
    ciphertext = [''] * key
    # pprint.pprint(ciphertext)

    # Loop through each column in ciphertext:
    for column in range(key):
        currentIndex = column

        # Keep looping until currentIndex goes past the message length:
        while currentIndex < len(message):
            # Place the character at currentIndex in message at the
            # end of the current column in the ciphertext list:
            ciphertext[column] += message[currentIndex]

            # Move currentIndex over:
            currentIndex += key

        # pprint.pprint(ciphertext)
    # Convert the ciphertext list into a single string value and return it:
    return ''.join(ciphertext)


def decryptMessage(key, message):
    # The transposition decrypt function will simulate the "columns" and
    # "rows" of the grid that the plaintext is written on by using a list
    # of strings. First, we need to calculate a few values.

    # The number of "columns" in our transposition grid:
    numOfColumns = int(math.ceil(len(message) / float(key)))
    # The number of "rows" in our grid:
    numOfRows = key
    # The number of "shaded boxes" in the last "column" of the grid:
    numOfShadedBoxes = (numOfColumns * numOfRows) - len(message)

    # Each string in plaintext represents a column in the grid:
    plaintext = [''] * numOfColumns
    # pprint.pprint(plaintext)

    # The column and row variables point to where in the grid the next
    # character in the encrypted message will go:
    column = 0
    row = 0

    for symbol in message:
        plaintext[column] += symbol
        column += 1  # Point to the next column.

        # If there are no more columns OR we're at a shaded box, go back
        # to the first column and the next row:
        if (
            column == numOfColumns or
                (
                    column == numOfColumns - 1 and
                    row >= numOfRows - numOfShadedBoxes
                    )
        ):

            column = 0
            row += 1

        # pprint.pprint(plaintext)

    return ''.join(plaintext)


def main(message):
    # myMessage = 'Common sense is not so common.'
    myMessage = message
    myKey = 8
    ciphertext = encryptMessage(myKey, myMessage)

    # Print the encrypted string in ciphertext to the screen, with
    # a | ("pipe" character) after it in case there are spaces at
    # the end of the encrypted message:
    print(ciphertext + '|')

    # pyperclip.copy()

    print(decryptMessage(myKey, ciphertext))


# If transpositionEncrypt.py is run (instead of imported as a module) call
# the main() function:
if __name__ == '__main__':
    plaintext = "Common sense is not so common."
    main(plaintext)

File: TranspositionCipherSolution.jl


function encrypt_transposition(string::String, key::Int=1)    
    # Each string in ciphertext represents a column in the grid:
    ciphertext = [[], [], [], [], [], [], [], []]
    array_words = []

    # Loop through each column in ciphertext:
    for column in 1:key
        currentIndex = column

        # Keep looping until currentIndex goes past the string length:
        while currentIndex <= length(string)
            # Place the character at currentIndex in string at the
            # end of the current column in the ciphertext list:
            #ciphertext[column] += string[currentIndex]
            append!(ciphertext[column], string[currentIndex])

            # Move currentIndex over:
            currentIndex += key
        end
    
    end

    println(ciphertext)
    println("Transpose:\n", permutedims(ciphertext))
    for array in ciphertext
        push!(array_words, join(array))
    end

    # Convert the ciphertext list into a single string value and return it:
    return join(array_words)
end


text = "Common sense is not so common."
println(encrypt_transposition(text, 8))
